File: number-of-islands/number-of-islands.java
class Solution {
    public int numIslands(char[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        int count = 0;
        for (int i = 0; i < m; i++){
            
            for(int j = 0 ; j < n; j++){
                
                if(grid[i][j] == '1'){
                    
                    if (helper(i,j, grid) == true){
                        count++;
                    }
                }
                
            }
        }
        return count;
    }
    
    public boolean helper(int i, int j, char[][] grid){
        int m = grid.length;
        int n = grid[0].length;
        if (i < 0 || i >= m || j < 0 || j >= n || grid[i][j] != '1') return false;
        
        grid[i][j] = 'a';
        
        helper(i+1, j, grid);
        helper(i-1, j, grid);
        helper(i,j+1, grid);
        helper(i, j-1, grid);
        
        return true;
        
        
    }
}

File: number-of-islands/number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        m = len(grid)
        n = len(grid[0])
        direc = [(0,1),(0,-1), (-1,0),(1,0)]
        
        
        
        def dfs(i,j):
            
            grid[i][j] = "0"
            
            for d in range(4):
                
                new_i = i + direc[d][0]
                new_j = j + direc[d][1]
                
                if new_i < 0 or new_i >= m or new_j < 0 or new_j >= n or grid[new_i][new_j] == "0":
                    continue
                dfs(new_i, new_j)
                
            
        num_islands = 0    
        for i in range(m):
            for j in range(n):
                if grid[i][j] != "0":
                    dfs(i,j)
                    num_islands += 1
        return num_islands
    